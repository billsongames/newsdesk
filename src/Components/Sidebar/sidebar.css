.sidebar-container {
  display: flex;
  flex-direction: row;
  background-color: var(--background);
  padding: 4px;

  height: 100%;
  left: 0;
  top: 0;
  z-index: 2;
}

.sidebar-buttons-container{
  display: flex;
  flex-flow: row;
  align-items: start;
}

.sidebar-filters-text{
    font-size: 14px;
    font-weight: bold;
    margin: 10px;
    color: var(--text);
    opacity: var(--opacityHigh);
}

.sidebar-filters-text > select {
  margin: 0px 0px 0px 6px;
}



.sidebar-button {
    background: none!important;
    border: none;
    font-family: 'Barlow', sans-serif;
    color: var(--text);
    text-decoration: none;
    cursor: pointer;
    font-size: 12px;
}

.sidebar-button:hover {
    text-decoration: underline;
}