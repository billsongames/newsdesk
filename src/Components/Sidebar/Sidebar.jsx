import React, { useState, useEffect } from "react";
import PropTypes from "prop-types"

import "./sidebar.css"

const Sidebar = (props) => {
  const { reverseOrder, articles, selected, setSelected } = props;
  const [newArticles, setNewArticles] = useState(true);
  const [oldArticles, setOldArticles] = useState(false);
  const [uniqueSources, setUniqueSources] = useState([])
  const [sourceSelected, setSourceSelected] = useState(selected)


  const newClickHandler = () => {
    setNewArticles(true);
    setOldArticles(false);
    reverseOrder();
  };

  const oldClickHandler = () => {
    setNewArticles(false);
    setOldArticles(true);
    reverseOrder();
  };

  useEffect(() => {
    setUniqueSources([...new Set(articles.map(article => article.source.name))])

  }, [articles])

//  const uniqueSources = [...new Set(articles.map(article => article.source.name))];

return (
  <div className="sidebar-container">
    <div className="sidebar-filters-text">
      Sort by Date:
      <button data-testid="button1" className="sidebar-button" onClick={newClickHandler} disabled={newArticles}>Most Recent</button>
      <button data-testid="button2" className="sidebar-button" onClick={oldClickHandler} disabled={oldArticles}>Oldest</button>
    </div>

    <div className="sidebar-filters-text">
      Filter by Source:

      <select value={selected}
        onChange={(e) => setSelected(e.target.value)}>
          console.log({selected})
        <option value="">-All Sources-</option>
          {uniqueSources.map((source) => (
          <option value={source} key={source}>{source}</option>
          ))}
      </select>
    </div>
  </div>
);
}

Sidebar.propTypes = {
  reverseOrder: PropTypes.func.isRequired,
  articles: PropTypes.array.isRequired,
  setSelected: PropTypes.func.isRequired
}

export default Sidebar;