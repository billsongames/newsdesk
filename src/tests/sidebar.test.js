import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";

import Sidebar from "../Components/Sidebar/Sidebar"

describe("Sidebar", () => {
    const validProps = {
        reverseOrder: jest.fn(),
        articles: [
            {
                title: "test_title_text",
                description: "test description text",
                content: "test description text",
                url: "test url",
                image: "test_image.jpg",
                publishedAt: "1970-01-01T00:0:00Z",
                source: {
                name: "test source",
                url: "test source url"
                }
            }
        ],
        selected: "",
        setSelected: jest.fn(),
    };

    it("renders the text and buttons to sort by date", () => {
        render(<Sidebar 
            reverseOrder={validProps.reverseOrder} 
            articles={validProps.articles} 
            selected={validProps.selected} 
            setSelected={validProps.setSelected} />);

        expect(screen.getByText("Sort by Date:")).toBeInTheDocument();
        expect(screen.getByText("Most Recent")).toBeInTheDocument();
        expect(screen.getByText("Oldest")).toBeInTheDocument();
    });

    it("calls the correct function when most recent clicked", async () => {
        render(<Sidebar 
            reverseOrder={validProps.reverseOrder} 
            articles={validProps.articles} 
            selected={validProps.selected} 
            setSelected={validProps.setSelected} />)
        const button = screen.getByTestId("button1");
        const button2 = screen.getByTestId("button2");
        await fireEvent.click(button2);
        await fireEvent.click(button);

        expect(validProps.reverseOrder).toHaveBeenCalled();
    });

    it("calls the correct function when oldest clicked", async () => {
        render(<Sidebar 
            reverseOrder={validProps.reverseOrder} 
            articles={validProps.articles} 
            selected={validProps.selected} 
            setSelected={validProps.setSelected} />)
        const button2 = screen.getByTestId("button2");
        await fireEvent.click(button2);

        expect(validProps.reverseOrder).toHaveBeenCalledTimes(1);
    });

    it("renders filters text", () => {
        render(<Sidebar 
            reverseOrder={validProps.reverseOrder} 
            articles={validProps.articles} 
            selected={validProps.selected} 
            setSelected={validProps.setSelected} />);

        expect(screen.getByText("Filter by Source:")).toBeInTheDocument();
    });

    xit("has the correct select and option elements", () => {
        render(<Sidebar 
            reverseOrder={validProps.reverseOrder} 
            articles={validProps.articles} 
            selected={validProps.selected} 
            setSelected={validProps.setSelected} />);
        });

    it("matches the snapshot", () => {
            const { asFragment } = render(<Sidebar 
                reverseOrder={validProps.reverseOrder} 
                articles={validProps.articles} 
                selected={validProps.selected} 
                setSelected={validProps.setSelected} />);

            expect(asFragment()).toMatchSnapshot();
    });
});
