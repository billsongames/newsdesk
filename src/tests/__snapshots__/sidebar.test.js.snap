// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sidebar-container"
  >
    <div
      class="sidebar-filters-text"
    >
      Sort by Date:
      <button
        class="sidebar-button"
        data-testid="button1"
        disabled=""
      >
        Most Recent
      </button>
      <button
        class="sidebar-button"
        data-testid="button2"
      >
        Oldest
      </button>
    </div>
    <div
      class="sidebar-filters-text"
    >
      Filter by Source:
      <select>
        console.log()
        <option
          value=""
        >
          -All Sources-
        </option>
        <option
          value="test source"
        >
          test source
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
